### PR TITLE
BFD-1876: New schema beneficiary related table migration and code

### PR DESCRIPTION
**Pull request by mackec1**
**JIRA Ticket:**
[BFD-1876](https://jira.cms.gov/browse/BFD-1876)

**User Story or Bug Summary:**

As part of the new table structure for BFD tables, this PR encapsulated work on Beneficiary-related tables; these are:

- beneficiaries
- beneficiary_monthly
- beneficiaries_history
- medicare_beneficiaryid_history
- skipped_rif_records

Each of the above tables will have its bene_id data type changed from varchar(15) to bigint.

BFD services (code) is updated to work vs. the new table name/structure.

### What Does This PR Do?

This PR includes 5 flyway (database) migration scripts:

1. migrate data from beneficiaries table to _beneficiaries_new_ table.
    - removed _bene_id_numeric_ column and any references to it.
3. migrate data from beneficiary_monthly to _beneficiary_monthly_new_ table.
4. migrate data from beneficiary_history to _beneficiary_history_new_ table.
5. migrate data from medicare_beneficiaryid_history to _medicare_beneficiaryid_history_new_ table.
6. migrate data from skipped_rif_records to _skipped_rif_records_new_ table.
7. The following worksheets are modified in the Excel spreadsheet to reflect changes to bene_id:
    - Beneficiary
    - Medicare Beneficiary Id
    - Beneficiary History
8. RifLayoutsProcessor.java is modified to generate appropriate entity classes that meet the _new table(s) structure.
    - removed temp code that handled _isFutureBigint_ as this is no longer needed.

All BFD server code, unit tests, and Integration tests modified to work with created _new_ tables.

### What Should Reviewers Watch For?

- Each Flyway script is well-structured and limited in scope to a single table.
- Java code is well structured and includes appropriate comments.
- Unit and IT tests can be successfully executred.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X] No

### What Needs to Be Merged and Deployed Before this PR?

This PR has no hard prerequisites; however, if a flyway script is introduced to master prior to this PR being merged, then the 5 flyways script will need to be renamed (re-numbered).

### Submitter Checklist

I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    10. There isn't too much of a burden on reviewers.
    11. Any problems it causes have a small "blast radius".
    12. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V80__new_schema_beneficiaries_history.sql [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                    bene_birth_dt,
+                    bene_sex_ident_cd,
+                    last_updated
+from
+    public.beneficiaries_history;
+    
+${logic.psql-only} alter table public.beneficiaries_history_new
+${logic.psql-only}    alter column bene_history_id     SET NOT NULL,
+${logic.psql-only}    alter column bene_id             SET NOT NULL,
+${logic.psql-only}    alter column hicn_unhashed       SET NOT NULL,
```
Comment (christrotteradhoc) [inline apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V80__new_schema_beneficiaries_history.sql:79]: Looks like this is currently nullable in prod

Comment (mackec1) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V80__new_schema_beneficiaries_history.sql:79]: fixed.

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V81__new_schema_skipped_rif_records.sql [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+${logic.psql-only}    alter column dml_ind             SET NOT NULL,
+${logic.psql-only}    alter column rif_data            SET NOT NULL;
+
+-- for PSQL need to define our primary key    
+${logic.psql-only} alter table public.skipped_rif_records_new
+${logic.psql-only}     add CONSTRAINT skipped_rif_records_new_pkey PRIMARY KEY (record_id);
+
+-- create an index on bene_id
+CREATE INDEX IF NOT EXISTS skipped_rif_records_new_bene_id_idx
+    ON public.skipped_rif_records_new (bene_id);
```
[the remark this answers is not in the document]
Comment (mackec1) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V81__new_schema_skipped_rif_records.sql:78]: not really; I whuile we do want to reinstate them, I assumed that would be part of the db 'cleanup' where we'll delete the orig bene tables, do renames from _new to orig table names, etc. Entirely up to you when this should be done...can re-instate them here, but we'll have to have _new in its definition, and then rename later.

[the remark this answers is not in the document]
Comment (mackec1) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V81__new_schema_skipped_rif_records.sql:78]: this article describes how we can do this without locking both FK tables:
https://travisofthenorth.com/blog/2017/2/2/postgres-adding-foreign-keys-with-zero-downtime

[the remark this answers is not in the document]
Comment (christrotteradhoc) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V81__new_schema_skipped_rif_records.sql:78]: given the non-trivial multi-step process I think we should wait so that we can test and plan for it appropriately

**Review by christrotteradhoc — APPROVED** [review]

**Review by christrotteradhoc — DISMISSED** [review]

**Review by cbrunefearless — APPROVED** [review]
lgtm

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/stu3/providers/LoadedFilterManagerIT.java [hunk 3]
```diff
@@ -77,7 +77,6 @@ public void emptyFilters() {
               assertFalse(filterManager.isResultSetEmpty(INVALID_BENE, testBound));
 
               // Refresh the filter list
-              filterManager.refreshFilters();
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/stu3/providers/LoadedFilterManagerIT.java:80]: why is this removed?

Comment (mackec1) [reply apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/stu3/providers/LoadedFilterManagerIT.java:80]: yea, it shouldn't have been; thx for catching that!

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2.java [hunk 4]
```diff
@@ -1579,12 +1579,24 @@ public static Bundle addResourcesToBundle(Bundle bundle, List<IBaseResource> res
...
 
+  /**
+   * @param beneId the {@link Long} beneficiary ID to log {@link
+   */
+  public static void logBeneIdToMdc(Long beneId) {
+    if (beneId != null) {
+      MDC.put("bene_id", String.valueOf(beneId));
+    }
+  }
+
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2.java:1599]: Can we just convert the original method to take Longs and keep it at a single method?

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/TransformerUtils.java [hunk 5]
```diff
@@ -3369,12 +3369,24 @@ public static Bundle addResourcesToBundle(Bundle bundle, List<IBaseResource> res
...
 
+  /**
+   * @param beneId the {@link Long} beneficiary ID to log {@link
+   */
+  public static void logBeneIdToMdc(Long beneId) {
+    if (beneId != null) {
+      MDC.put("bene_id", String.valueOf(beneId));
+    }
+  }
+
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/TransformerUtils.java:3389]: same comment as above. and actually this is a shame that we have all of this twice but we are doing some work early next sprint to fix some other issues with this bene_id logging so we will sort that out then so I don't think it should become scope creep on this PR.

File: apps/bfd-model/bfd-model-rif/src/main/java/gov/cms/bfd/model/rif/LoadedBatch.java [hunk 6]
```diff
@@ -195,10 +194,18 @@ public static LoadedBatch combine(LoadedBatch a, LoadedBatch b) {
...
+      return "";
+    }
+    StringBuilder sb = new StringBuilder();
+    for (int i = 0; i < list.size(); i++) {
+      if (i > 0) {
+        sb.append(SEPARATOR);
+      }
+      sb.append(String.valueOf(list.get(i)));
+    }
+    return sb.toString();
```
[the remark this answers is not in the document]
Comment (christrotteradhoc) [reply apps/bfd-model/bfd-model-rif/src/main/java/gov/cms/bfd/model/rif/LoadedBatch.java:208]: My comment is less about performance (I think both will perform similarly) and more about simplicity.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2.java (hunk 4 above)
Comment (christrotteradhoc) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2.java:1599]: I would not be surprised if the JVM optimizes these to the same thing. Introducing code duplication to save on a single allocation per call when there are any thousands of allocations in HAPI seems like premature optimization to me.

File: apps/bfd-model/bfd-model-rif/src/main/java/gov/cms/bfd/model/rif/LoadedBatch.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (christrotteradhoc) [reply apps/bfd-model/bfd-model-rif/src/main/java/gov/cms/bfd/model/rif/LoadedBatch.java:208]: I think we'll convert to an array and varargs signature in the upcoming change anyway so doesn't matter much as long as it works

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2.java (hunk 4 above)
Comment (christrotteradhoc) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2.java:1599]: I think we'll convert to an array and varargs signature in the upcoming change anyway so doesn't matter much as long as it works

File: apps/bfd-model/bfd-model-rif/src/main/java/gov/cms/bfd/model/rif/LoadedBatch.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (mackec1) [reply apps/bfd-model/bfd-model-rif/src/main/java/gov/cms/bfd/model/rif/LoadedBatch.java:208]: sounds good...less code lines makes sense here. Will change.

**Review by christrotteradhoc — DISMISSED** [review]

**Review by christrotteradhoc — APPROVED** [review]
Great work! But do we have to be reminded of those hanging chads?

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V80__new_schema_beneficiaries.sql [hunk 7]
```diff
@@ -0,0 +1,727 @@
+-- NEW_SCHEMA_BENEFICIARIES.SQL
+-- flyway migration for BENEFICIARIES table into
+-- a new schema structure that:
+--   1) changes data type of BENE_ID from varchar to BIGINT.
+--
+-- Once current table data is migrated to new table name/structure, a 
+-- subsequent PR will be deployed that changes the ORM model(s) and
+-- operational code for Beneficiaries table.
```
Comment (lsmitchell) [inline apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V80__new_schema_beneficiaries.sql:8]: I think this isn't true anymore right? Since this is a combo code/migration PR now that we have the migrator.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4CoverageResourceProvider.java [hunk 8]
```diff
@@ -187,22 +194,22 @@ public Bundle searchByBeneficiary(
...
-    TransformerUtilsV2.logBeneIdToMdc(Arrays.asList(beneficiary.getIdPart()));
+    TransformerUtilsV2.logBeneIdToMdc(beneficiaryId);
 
     return TransformerUtilsV2.createBundle(
         paging, coverages, loadedFilterManager.getTransactionTime());
   }
 
   /**
-   * @param beneficiaryId the {@link Beneficiary#getBeneficiaryId()} value to find a matching {@link
+   * @param beneIdStr the {@link Beneficiary#getBeneficiaryId()} value to find a matching {@link
```
Comment (lsmitchell) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4CoverageResourceProvider.java:204]: Error on the param name in the doc here

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2.java [hunk 9]
```diff
@@ -1579,12 +1579,20 @@ public static Bundle addResourcesToBundle(Bundle bundle, List<IBaseResource> res
     return bundle;
   }
 
+  /** @param beneIds the {@link Collection} beneficiary IDs to log */
```
Comment (lsmitchell) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2.java:1582]: Missing description on this and the one below (one day I'll get checkstyle enabled for the rest of these...)

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/CoverageResourceProvider.java [hunk 10]
```diff
@@ -187,33 +187,33 @@ public Bundle searchByBeneficiary(
...
-    TransformerUtils.logBeneIdToMdc(Arrays.asList(beneficiary.getIdPart()));
+    TransformerUtils.logBeneIdToMdc(beneficiaryId);
 
     return TransformerUtils.createBundle(
         paging, coverages, loadedFilterManager.getTransactionTime());
   }
 
   /**
-   * @param beneficiaryId the {@link Beneficiary#getBeneficiaryId()} value to find a matching {@link
+   * @param beneIdStr the {@link Beneficiary#getBeneficiaryId()} value to find a matching {@link
```
Comment (lsmitchell) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/CoverageResourceProvider.java:197]: param name error again

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/TransformerUtils.java [hunk 11]
```diff
@@ -3369,12 +3369,20 @@ public static Bundle addResourcesToBundle(Bundle bundle, List<IBaseResource> res
     return bundle;
   }
 
+  /** @param beneIds the {@link Collection} of beneficiary IDs top log */
```
Comment (lsmitchell) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/TransformerUtils.java:3372]: Missing doc description for these two new docs as well.

**Review by christrotteradhoc — APPROVED** [review]
Changes since my last approval lood good ; re-approving.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4CoverageResourceProvider.java (hunk 8 above)
Comment (mackec1) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4CoverageResourceProvider.java]: done

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2.java (hunk 9 above)
Comment (mackec1) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2.java]: done

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/CoverageResourceProvider.java (hunk 10 above)
Comment (mackec1) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/CoverageResourceProvider.java:197]: done

**Review by JeremyIR — APPROVED** [review]
